x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/SemanticTextInfoExtractor.java

```diff
@@ -33,15 +33,11 @@ public static Set<String> extractIndexesReferencingInferenceEndpoints(Metadata m
         Map<String, IndexMetadata> indices = metadata.indices();
 
         indices.forEach((indexName, indexMetadata) -> {
-            if (indexMetadata.getInferenceFields() != null) {
-                Map<String, InferenceFieldMetadata> inferenceFields = indexMetadata.getInferenceFields();
-                if (inferenceFields.entrySet()
-                    .stream()
-                    .anyMatch(
-                        entry -> entry.getValue().getInferenceId() != null && endpointIds.contains(entry.getValue().getInferenceId())
-                    )) {
-                    referenceIndices.add(indexName);
-                }
+            Map<String, InferenceFieldMetadata> inferenceFields = indexMetadata.getInferenceFields();
+            if (inferenceFields.values()
+                .stream()
+                .anyMatch(im -> endpointIds.contains(im.getInferenceId()) || endpointIds.contains(im.getSearchInferenceId()))) {
+                referenceIndices.add(indexName);
```

Comment:
I simplified this logic because `indexMetadata.getInferenceFields()`, `im.getInferenceId()`, and `im.getSearchInferenceId()` will never return null

Reply:
How do we know they will never return null?

Reply:
For `indexMetadata.getInferenceFields()`, we know this because the only place where a [new `IndexMetadata` object is constructed](https://github.com/Mikep86/elasticsearch/blob/48f790c8d66abbaec315fb639981691dd47e2c22/server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java#L2406-L2406) (besides copy constructors) explicitly passes a non-null `InferenceFieldMetadata` map. You can also look at how `inferenceFields` is used in `IndexMetadata` ([example](https://github.com/Mikep86/elasticsearch/blob/48f790c8d66abbaec315fb639981691dd47e2c22/server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java#L1467-L1467), [example](https://github.com/Mikep86/elasticsearch/blob/48f790c8d66abbaec315fb639981691dd47e2c22/server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java#L1490-L1490), [example](https://github.com/Mikep86/elasticsearch/blob/48f790c8d66abbaec315fb639981691dd47e2c22/server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java#L1860-L1860)), it's clear from that it can never be null or we would get NPEs on internal `IndexMetadata` operations. Finally, [`inferenceFields` is not annotated as `@Nullable` in the `IndexMetadata` constructor](https://github.com/Mikep86/elasticsearch/blob/48f790c8d66abbaec315fb639981691dd47e2c22/server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java#L663-L663). Normally, this wouldn't mean much, but in `IndexMetadata` we are particularly careful about these things because an oversight in this class means a CrowdStrike-style distributed production outage 🫠.

For `InferenceFieldMetadata`, we know that `getInferenceId()` and `getSearchInferenceId()` will never return null because they are [specifically checked in the constructor to be non-null](https://github.com/Mikep86/elasticsearch/blob/48f790c8d66abbaec315fb639981691dd47e2c22/server/src/main/java/org/elasticsearch/cluster/metadata/InferenceFieldMetadata.java#L51-L52).